src/filesystem.c

```diff
@@ -248,45 +248,182 @@ rcutils_calculate_directory_size(const char * directory_path, rcutils_allocator_
       "Path is not a directory: %s\n", directory_path);
     return dir_size;
   }
+
+  return rcutils_calculate_directory_size_with_recursion(directory_path, 1, allocator);
+}
+
+typedef struct dir_list_t
+{
+  char * path;
+  uint32_t depth;  // Compare with base path
+  struct dir_list_t * next;
+} dir_list_t;
+
+static inline void free_dir_list(dir_list_t * dir_list, rcutils_allocator_t allocator)
+{
+  dir_list_t * next_dir;
+  do {
+    next_dir = dir_list->next;
+    allocator.deallocate(dir_list->path, allocator.state);
+    allocator.deallocate(dir_list, allocator.state);
+    dir_list = next_dir;
+  } while (dir_list);
+}
+
+size_t
+rcutils_calculate_directory_size_with_recursion(
+  const char * directory_path,
+  const uint32_t max_deepth,
+  rcutils_allocator_t allocator)
+{
+  dir_list_t * dir_list = NULL;
+  size_t dir_size = 0;
+
+  dir_list = (dir_list_t *) allocator.allocate(sizeof(dir_list_t), allocator.state);
+  if (NULL == dir_list) {
+    RCUTILS_SAFE_FWRITE_TO_STDERR("Failed to allocate memory !\n");
+    return 0;
+  }
+
 #ifdef _WIN32
-  char * path = rcutils_join_path(directory_path, "*", allocator);
-  WIN32_FIND_DATA data;
-  HANDLE handle = FindFirstFile(path, &data);
-  if (INVALID_HANDLE_VALUE == handle) {
-    RCUTILS_SAFE_FWRITE_TO_STDERR_WITH_FORMAT_STRING(
-      "Can't open directory %s. Error code: %lu\n", directory_path, GetLastError());
-    return dir_size;
+  HANDLE handle = INVALID_HANDLE_VALUE;
+
+  dir_list->next = NULL;
+  dir_list->depth = 1;
+  dir_list->path = rcutils_join_path(directory_path, "*", allocator);
+  if (NULL == dir_list->path) {
+    RCUTILS_SAFE_FWRITE_TO_STDERR("Failed to duplicate directory path !\n");
+    goto fail;
   }
-  allocator.deallocate(path, allocator.state);
-  if (handle != INVALID_HANDLE_VALUE) {
+
+  do {
+    WIN32_FIND_DATA data;
+    handle = FindFirstFile(dir_list->path, &data);
+    if (INVALID_HANDLE_VALUE == handle) {
+      RCUTILS_SAFE_FWRITE_TO_STDERR_WITH_FORMAT_STRING(
+        "Can't open directory %s. Error code: %lu\n", dir_list->path, GetLastError());
+      goto fail;
+    }
+
     do {
       // Skip over local folder handle (`.`) and parent folder (`..`)
       if (strcmp(data.cFileName, ".") != 0 && strcmp(data.cFileName, "..") != 0) {
```

Comment:
I think the loop would be a little easier to read if we `continue` early on `.` and `..`.  Something like:

```
  do {
    if (cFilename == "." || cFilename == "..") {
      continue;
    }
    ...
  } while (FindNextFile(handle, &data));
```